### DISCUSSION TITLE
[OBA 0] Требования к движку

### DISCUSSION THREAD

**Discussion by nergal-perm** [nergal-perm/tg-core#30] [Текущие задачи]
Здесь будем предлагать и обсуждать набор требований к движку настольных игр на этапе предварительного анализа в рамках задачи #28. Напоминаю, что требование - это одна конкретная вещь, которую должна делать (или должна не делать) система.

Можно попробовать в рамках этого обсуждения каждое требование записывать в виде отдельного комментария, и вести обсуждение в ветке этого комментария (с помощью "ответов" на комментарий). Когда в ходе обсуждения требования решим, что требование можно зафиксировать - поставим реакцию "палец вверх" и перенесем требование на страницу wiki.

**Comment by nergal-perm** [comment]
Движок должен отслеживать общедоступное состояние игры (например, состояние игрового стола)

**Comment by nergal-perm** [comment]
Движок должен отслеживать закрытые состояния игры (например, карты в руке у игроков)

**Comment by nergal-perm** [comment]
### Движок должен отслеживать фазы игры, например, "ход игрока 1"

Мне даже подумалось, что каждая Игра может в момент инициализации выстраивать свою структуру, определяя фазы, раунды и прочие переходы хода. При этом для каждой такой структурной сущности можно будет определять различные условия, например, условия окончания игры, как в начале, так и в конце. Более того, у некоторых игр на разных этапах изменяются условия, например, в "Энергосети" есть три этапа, на которых рынок ресурсов и рынок электростанций ведут себя по-разному.

В общем, это, наверное, одно из центральных требований к движку, потому что он должен быть заточен именно под пошаговые игры.

**Reply by nergal-perm** [reply]
Есть у меня, правда, опасение, что может получиться так, что выделить общие структуры ходов или механизм, который позволит обобщенно управлять очередностью действий игроков, у нас не получится, потому что отличий в механике конкретных игр будет гораздо больше, чем схожих моментов. Так что это требование нужно проработать в первую очередь, возможно, все видение движка изменится на основании результатов проработки.

**Reply by citerm** [reply]
Мне сложно выделить новую сущность движок и предъявлять к нему требования. Ты уверен, что кроме низкоуровневого обслуживания подсистем "Состояния", "Способности" и "Правила" у движка есть еще какие то общие требования. 
Инициализация игры (конкретные действия по первоначальной настройке игры) определена в правилах, в чем здесь роль движка? Все уникальные требования для каждой игры определены в правилах, движок на мой взгляд только лишь осуществляет стандартизированное взаимодействие между подсистемами.

**Comment by nergal-perm** [comment]
Движок должен передавать игрокам списки допустимых действий

**Comment by nergal-perm** [comment]
Движок должен сохранять историю действий игроков

**Comment by nergal-perm** [comment]
Движок должен подключать игроков к игре, если это возможно

**Comment by nergal-perm** [comment]
Движок должен отключать игроков от игры

**Comment by nergal-perm** [comment]
Движок не должен обрабатывать недопустимые действия игроков
